Updated for newer versions of ejabberd

malexmave opened the pull request:
As [promised](https://twitter.com/malexmave/status/550763597828485120), I updated the guide for the current version of ejabberd, which uses a different config file syntax. I also added some additional hints (specifying cipher orders and enforcing them, changing the logging options, registration settings and password storage).

Feedback is obviously welcome, but what really **needs** to be reviewed before merging is the cipher specs [here](https://github.com/malexmave/Applied-Crypto-Hardening/blob/ejabberd-update/src/configuration/IM/ejabberd/ejabberd.yml#L139) and [here](https://github.com/malexmave/Applied-Crypto-Hardening/blob/ejabberd-update/src/configuration/IM/ejabberd/ejabberd.yml#L211) (the cipher strings are identical, but not reviewed and may contain bad ciphers. They resulted from me trying to achieve a cipher order which favored ciphers with PFS and forbade terrible ciphers like DES and RC4). Please provide me with a better cipher string than the one I provided, and I will add it (keep in mind that some clients only support comparatively terrible ciphers, as per [this article](https://blog.thijsalkema.de/me/blog//blog/2013/09/02/the-state-of-tls-on-xmpp-3/)).

I also removed the outdated config for older versions of ejabberd (it was called ejabberd.cfg instead of ejabberd.yml and used the erlang format instead of YAML) in order to avoid confusion if people look into the directory.

This is my first time contributing to this guide, so please let me know about any formatting problems or similar errors I may have introduced, or guidelines I may have violated. Thank you for your great work and time.

(sorry for the large commit, but config files tend to be rather large... :))

sebix commented on src/practical_settings/im.tex:
```diff
@@ -31,30 +31,46 @@ \subsection{ejabberd}
+Other options you may want to take a look at are:
+\begin{itemize}
+\item The log level logs all login, including username and IP address, by default. Change the \texttt{loglevel} setting to avoid this:\\
+\configfile{ejabberd.yml}{31-39}{Changing the logging behaviour}
```
Isn't this out of scope (as some of the following recommendations?) The defined scope is Crypto - not general security and privacy. These recommendations better fit in the default (and shipped) config.

malexmave replied:
I opted to include it because I believe in minimizing the amount of data an attacker can get if he/she compromises the server, but I'll leave it to the editors to decide if they want to have this included or not. 

The same goes for the open registration: The default may be a bit counterintuitive, so I'd rather err on the side of caution and mention it, but if the maintainers / editors want it out, I'll take it out.

sebix replied:
While this may be true, it is nevertheless not in the scope of this document. Minimizing damage by attackers has very little to do with securing the transport paths using state of the art crypto. If we would do this for all covered programs, the entire section about practical settings would be about two times bigger.

Disclaimer: I´m not a maintainer.

azet replied:
ACK.

malexmave commented:
I changed the cipher string to `EDH+CAMELLIA:EDH+aRSA:EECDH+aRSA+AESGCM:EECDH+aRSA+SHA256:EECDH:+CAMELLIA128:+AES128:+SSLv3:!aNULL:!eNULL:!LOW:!3DES:!MD5:!EXP:!PSK:!DSS:!RC4:!SEED:!IDEA:!ECDSA:kEDH:CAMELLIA128-SHA:AES128-SHA` as per the second recommended cipher list of the guide. I did not put in the first because of the [known bad cipher coverage on some clients](https://blog.thijsalkema.de/me/blog//blog/2013/09/02/the-state-of-tls-on-xmpp-3/) and the possible limitations in s2s-connections that may entail.

azet commented:
> I changed the cipher string to EDH+CAMELLIA:EDH+aRSA:EECDH+aRSA+AESGCM:EECDH+aRSA+SHA256:EECDH:+CAMELLIA128:+AES128:+SSLv3:!aNULL:!eNULL:!LOW:!3DES:!MD5:!EXP:!PSK:!DSS:!RC4:!SEED:!IDEA:!ECDSA:kEDH:CAMELLIA128-SHA:AES128-SHA as per the second recommended cipher list of the guide. I did not put in the first because of the known bad cipher coverage on some clients and the possible limitations in s2s-connections that may entail.

We've updated our cipherstrings (I know the PDF on the webpage hasn't been updated in quite a while), see git sources: https://git.bettercrypto.org/ach-master.git/blob/HEAD:/src/common/cipherStringB.tex

Please use this one.

As for 'updating': I'd recommend to add another 'version' of the same configuration, leave the existing configuration for the older `ejabberd` and add a new one (as with other daemons, see openssh et cetera) mentioning the version you've tested and added to the guide.

**It is mandatory that additions be tested (ideally on different distributions) beforehand.**

We currently have no testing infrastructure in place, but might switch to using `rspec` or something similar in the future. For this, though, we need more people actively contributing to the project - since there are only a few of the main authors left that are actively contributing, most of which have very limited time.

malexmave commented:
Is the active one (the one that is not commented out) from your link the new one? Because, from the comments, it seems to be, and that's the one I am using...

I'll put the old ejabberd version back in, following the example of OpenSSH.

As for testing, I don't really have a setup for that ready, either. I'll see what I can do. Other testers are obviously welcome (mine was tested on Ubuntu 12.04.1 LTS).

azet commented:
@malexmave:
> Is the active one (the one that is not commented out) from your link the new one? Because, from the comments, it seems to be, and that's the one I am using...
Oh, yes. You're right.

> I'll put the old ejabberd version back in, following the example of OpenSSH.
Thanks!

> As for testing, I don't really have a setup for that ready, either. I'll see what I can do. Other testers are obviously welcome (mine was tested on Ubuntu 12.04.1 LTS).
Which means it'll probably work for Debian wheezy. But that's not guaranteed. If you haven't tested any other OS or distro. just state the one you've used.

Thanks for your addition!

malexmave commented:
I'll set up a VM with an up-to-date debian tomorrow and see if it runs with the config file, create an account and see if I can connect to it. Can't really do much more checking, but the majority of errors should occur with those steps.

Other than that, as you can see I pushed the old stuff back in. A quick glance if it meets the formatting guidelines would be appreciated.

azet commented:
Thanks! Looks good. Other reviews very welcome.

malexmave commented:
Just tested it with a (self-compiled) version 14.12 on the latest debian (Debian 7 x64). As per [this comment thread](https://github.com/BetterCrypto/Applied-Crypto-Hardening/pull/90#discussion_r22645787), I also removed the off-topic information from the guide.

azet commented:
Thanks. I'll need a second person to review. For me this PR looks good so far. It's too large for a single person though.

schwindp commented:
You might want to add the info (for existing servers with ODBC enabled) that after enabling hashed passwords, you need to run `ejabberd_auth_odbc:convert_to_scram(<<"your.server.name">>` inside an ejabberdctl debug shell to have them automagically converted.

Looking good to me though.

malexmave commented:
Thanks for the info, will fix that as soon as I get the chance.

Also, to quote myself from the mailing list:

> While testing my ejabberd for the LogJam vulnerability, I noticed that my ejabberd negotiates 1024 bit keys. The current cipherstring is `DH+CAMELLIA:EDH+aRSA:EECDH+aRSA+AESGCM:EECDH+aRSA+SHA256:EECDH:+CAMELLIA128:+AES128:+SSLv3:!aNULL:!eNULL:!LOW:!3DES:!MD5:!EXP:!PSK:!DSS:!RC4:!SEED:!IDEA:!ECDSA:kEDH:CAMELLIA128-SHA:AES128-SHA`, as per the recommendations in the Guide (Cipher String B, page 67), and the same cipher string is used in the config in the PR.

> I am not quite clear on whether this is the fault of the cipher string or the implementation of ejabberd. As far as I know, ejabberd uses openssl, which properly negotiates 4096 bit keys for my apache web server with a slightly different cipher string, namely: `EECDH+ECDSA+AESGCM EECDH+aRSA+AESGCM EECDH+ECDSA+SHA384  EECDH+ECDSA+SHA256 EECDH+aRSA+SHA384 EECDH+aRSA+SHA256 EECDH+aRSA+RC4 EECDH EDH+aRSA RC4 !aNULL !eNULL !LOW !3DES !MD5 !EXP !PSK !SRP !DSS !RC4`

> I'd like to fix this problem, both for my own server and for the config in the pull request. Does anyone have an idea what is going on here and what would have to be changed to get ejabberd to use better DH keys?

If anyone has any input on this, it would be appreciated, since I'd rather not provide the guide with a config that negotiates small keys, if at all possible.

ignisf commented:
See https://github.com/processone/tls/pull/11
Saw it here: https://twitter.com/jabbercccde/status/603280498899681281

malexmave commented:
Okay, we should probably hold off on merging this PR until that is settled and I had a chance to update the guide. Thanks for the information, that clears things up.

azet commented:
IIRC Apache has special handling to establish DH params relative to the strength of the symmetric cipher configured/negotiated. I'm not so sure that OpenSSL does that in general.

PR looks fine, but let's hold off for now until this has been resolved.

sebix commented:
Any updates on this issue @malexmave?

malexmave commented:
I'm currently in the middle of writing my master thesis, so I've been pretty swamped. I'll try to get this done ASAP (Ironically, I may find time when I am at 32c3, since I will not even try to work for my thesis there). So, it's still on my radar and I'll try to get it done before the year is out.

malexmave commented:
Updated the PR with the information on DH parameters. Sorry for taking so long.

As I removed the section about scrambled password storage, I did not incorporate [the suggestion by](https://github.com/BetterCrypto/Applied-Crypto-Hardening/pull/90#issuecomment-105154387) @schwindp. From my perspective, the PR should be finished now, reviews welcome.

sebix commented on src/configuration/IM/ejabberd/14.12/ejabberd.yml:
```diff
@@ -0,0 +1,694 @@
+  # of the other server
+
+s2s_ciphers: "EDH+CAMELLIA:EDH+aRSA:EECDH+aRSA+AESGCM:EECDH+aRSA+SHA256:EECDH:+CAMELLIA128:+AES128:+SSLv3:!aNULL:!eNULL:!LOW:!3DES:!MD5:!EXP:!PSK:!DSS:!RC4:!SEED:!IDEA:!ECDSA:kEDH:CAMELLIA128-SHA:AES128-SHA"
+# TODO Is this cipher order fine?
```
TODO can be removed, as the string is the same as in `src/common/cipherStringB.tex`.

sebix commented on src/practical_settings/im.tex:
```diff
@@ -31,30 +31,55 @@ \subsection{ejabberd}
 \subsubsection{Tested with Versions}
 \begin{itemize*}
-  \item Debian Wheezy 2.1.10-4+deb7u1
+  \item 14.12 (Ubuntu 14.04.1 LTS x64, Debian 7 x64)
```
Should be as in other sections:

```latex
\begin{itemize*}
  \item ejabberd 14.12, Debian 7 Wheezy
  \item ejabberd 14.12, Ubuntu 14.04 Trusty
\end{itemize*}
```
We do not honor the architecture.

sebix commented on src/practical_settings/im.tex:
```diff
@@ -31,30 +31,55 @@ \subsection{ejabberd}
+It is possible to explicitly specify a cipher string for TLS connections.
+\configfile{14.12/ejabberd.yml}{104-107,113-114,119-119,123-123,125-125,127-127,134-135,138-138,142-142,144-146,195-195,203-203,207-207,210-213,217-217}{Specifying a cipher order and enforcing it}
+
+Note that we are setting the SSL option cipher\_server\_preference. This enforces our cipher order when negotiating which ciphers are used, as the cipher order of some clients chooses weak ciphers over stronger ciphers.\footnote{\url{https://blog.thijsalkema.de/me/blog//blog/2013/09/02/the-state-of-tls-on-xmpp-3/}}
```
Use ``\texttt{cipher\_server\_preference}`` to format code inline.

malexmave commented:
Thanks @sebix, I will implement these changes this evening.

malexmave commented:
Done. Let me know if I missed anything else.

malexmave commented:
I just noticed that even though I specified a DH parameter file, the server still uses 1024 bit DH groups for the key negotiation. Please hold off on merging this until I can figure out what went wrong and potentially fix the configuration file in this PR.

malexmave commented:
Just verified that after an update to ejabberd version 16.01, it definitely works as intended. I have no idea if my changes to indicate this are acceptable - please review and advise how I should change it, if needed.

malexmave commented:
Just added another small change, as the wording in the section about the older versions was confusing.

malexmave commented:
Bump. This could still use some reviews and then a merge, before it is outdated again :)

azet commented:
Looks good to me, everything seems to have been addressed. Merging.

@sebix - just to be sure, please take a look as well. maybe @MacLemon could also help out? He'd some experience with `ejabberd` as far as I remember.

sebix commented:
Sorry, I don't have any experience with ejabberd.
